src/System.Net.Http/src/System/Net/Http/Managed/HttpEnvironmentProxy.cs

```diff
@@ -0,0 +1,260 @@
+        private ICredentials _credientials;
+
+        public static HttpEnvironmentProxy TryToCreate()
+        {
+            // Get environmental variables. Protocol specific take precedence over
+            // general all_*, lower case variable has precedence over upper case.
+            // Note that curl uses HTTPS_PROXY but not HTTP_PROXY.
+            // For http, only http_proxy and generic variables are used.
+
+            Uri httpProxy = GetUriFromString(Environment.GetEnvironmentVariable(EnvHttpProxyLC));
+            Uri httpsProxy = GetUriFromString(Environment.GetEnvironmentVariable(EnvHttpsProxyLC)) ??
+                             GetUriFromString(Environment.GetEnvironmentVariable(EnvHttpsProxyUC));
+
+            if (httpProxy == null || httpsProxy == null)
+            {
+                Uri allProxy = GetUriFromString(Environment.GetEnvironmentVariable(EnvAllProxyLC)) ??
+                                GetUriFromString(Environment.GetEnvironmentVariable(EnvAllProxyUC));
+
+                if (httpProxy == null)
+                {
+                    httpProxy = allProxy;
+                }
+                if (httpsProxy == null)
+                {
+                    httpsProxy = allProxy;
+                }
+            }
+
+            // Do not instantiate if nothing is set.
+            // Caller may pick some other proxy type.
+            if (httpProxy == null && httpsProxy == null)
+            {
+                return null;
+            }
+
+            return new HttpEnvironmentProxy(httpProxy, httpsProxy, Environment.GetEnvironmentVariable(EnvNoProxyLC));
+        }
+
+        private HttpEnvironmentProxy(Uri httpProxy, Uri httpsProxy, string bypassList)
+        {
+            NetworkCredential httpCred = null;
+            NetworkCredential httpsCred = null;
+            bool sameAuth = false;
+
+            if (httpProxy != null)
+            {
+                _http = httpProxy;
+                httpCred = GetCredientialsFromString(_http.UserInfo);
+            }
+            if (httpsProxy != null)
+            {
+                _https = httpsProxy;
+                httpsCred = GetCredientialsFromString(_https.UserInfo);
+                if ((_http != null) && (_http.UserInfo == _https.UserInfo))
+                {
+                    sameAuth = true;
+                }
+            }
+            if (httpCred != null || httpsCred != null)
+            {
+                // If only one protocol is set or both protocols have same auth,
+                // use standard credential object
+                if (sameAuth || httpCred == null || httpsCred == null)
+                {
+                    _credientials = httpCred ?? httpsCred;
+                }
+                else
+                {
+                    // use wrapper so we can decide later based on uri
+                    _credientials = new HttpEnvironmentProxyCredientials(httpCred, httpsCred);
```

Comment:
Why not just always use the wrapper?  Makes the code less complicated to do so, and makes the behavior consistent.

Reply:
I was thinking about it but than of each use we will have check for the Scheme in GetCredential(). If you think that is ok I can change it. I agree that it would simplify the code.

Reply:
This is related to my comment above:

> The only Uris that should be valid here are literally the two Uris you return from GetProxy. So it's probably better to compare the passed uri to exactly those uris. For one, it will almost always be the same Uri instance, so the == check is cheap. For another, if for some bizarre reason a user passes some other Uri here, it's better to return null.

If you change to compare Uris (instead of scheme) as I suggest above, then the cost should be very cheap, and we can just always use the wrapper.